torch/_dynamo/variables/dicts.py

```diff
@@ -1,45 +1,167 @@
+
+
+def is_hashable_python_var(x):
+    from torch import Tensor
+
+    # Note: Keep me in sync with is_hashable!
+    # Even better, we should have a map of functions connecting the two
+    from ..allowed_functions import is_builtin_callable
+
+    return (
+        ConstantVariable.is_literal(x)
+        or isinstance(x, (Tensor, enum.Enum))
+        or is_builtin_callable(x)
+        or (isinstance(x, tuple) and all(is_hashable_python_var(e) for e in x))
+    )
+
+
+def is_hashable(x):
+    # Keep me in sync with is_hashable_python_var!
+    # Even better, we should have a map of functions connecting the two
+    if isinstance(x, variables.TensorVariable):
+        # Tensors are hashable if they have an example_value (a fake tensor)
+        # Most VT's should have one.
+        # It'd be nice if at some point we could assert that they all have one
+        return x.as_proxy().node.meta.get("example_value") is not None
+    elif isinstance(x, variables.TupleVariable):
+        return all(is_hashable(e) for e in x.items)
+    else:
+        return isinstance(
+            x,
+            (
+                variables.BuiltinVariable,
+                variables.SymNodeVariable,
+                variables.ConstantVariable,
+                variables.EnumVariable,
+            ),
+        )
 
 
 class ConstDictVariable(VariableTracker):
-    def __init__(self, items, user_cls, **kwargs):
+    class _HashableTracker:
+        """
+        Auxiliary opaque internal class that wraps a VariableTracker and makes it hashable
+        This should not be seen or touched by anything outside of ConstDictVariable and its children
+        Note that it's also fine to put VTs into dictionaries and sets, but doing so does not take into account aliasing
+        """
+
+        def __init__(self, vt):
+            # We specialize SymNodes
+            vt = specialize_symnode(vt)
+            assert is_hashable(vt), type(vt)
+            self.vt = vt
+
+        @property
+        def underlying_value(self):
+            if isinstance(self.vt, variables.TensorVariable):
+                x = self.vt.as_proxy().node.meta["example_value"]
+            elif isinstance(self.vt, variables.TupleVariable):
+                Hashable = ConstDictVariable._HashableTracker
+                x = tuple(Hashable(e).underlying_value for e in self.vt.items)
+            else:
+                x = self.vt.as_python_constant()
+            return x
+
+        def __hash__(self):
+            return hash(self.underlying_value)
+
+        @staticmethod
+        def _eq_impl(a: VariableTracker, b: VariableTracker):
+            # TODO: Put this in utils and share it between variables/builtin.py and here
+            if type(a) != type(b):
+                return False
+            elif isinstance(a, tuple):
+                return len(a) == len(b) and all(_eq_impl(u, v) for u, v in zip(a, b))
+            elif is_fake(a):
```

Comment:
If `a` is a VariableTracker, when will this be true?

Reply:
Sorry, the types annotations here are incorrect. The inputs are python objects.